File: pe-core/src/main/java/org/saar/core/renderer/r2d/RenderNode2D.java
package org.saar.core.renderer.r2d;

import org.saar.core.node.AbstractNode;
import org.saar.core.node.RenderNode;

public class RenderNode2D extends AbstractNode implements RenderNode {

    private final ModelBuffers2D buffers;

    public RenderNode2D(Vertex2D[] vertices, int[] indices) {
        this.buffers = new ModelBuffers2DOneVbo(vertices.length, indices.length);
        this.buffers.load(vertices, indices);
    }

    @Override
    public void render() {
        this.buffers.getModel().draw();
    }

    @Override
    public void delete() {
        this.buffers.getModel().delete();
    }
}

File: pe-core/src/main/kotlin/org/saar/core/renderer/p2d/RenderNode2D.kt
package org.saar.core.renderer.p2d

import org.saar.core.node.Node
import org.saar.core.renderer.RenderContext
import org.saar.core.renderer.RenderPass

interface RenderNode2D : Node {
    fun render2D(context: RenderContext)
}

fun RenderNode2D.asRenderNode2D() = object : RenderPass {
    override fun render(context: RenderContext) = this@asRenderNode2D.render2D(context)

    override fun delete() = this@asRenderNode2D.delete()
}